src/drivers/intel/cavs/ssp.c

```diff
@@ -962,6 +963,177 @@ static int ssp_get_fifo(struct dai *dai, int direction, int stream_id)
+#endif
+}
+
+static inline uint32_t ssp_ts_local_walclk_addr(int index)
+{
+#if CONFIG_APOLLOLAKE
+	if (index < DAI_NUM_SSP_BASE)
+		return TIMESTAMP_BASE + TS_I2S_LOCAL_WALCLK(index);
+	else
+		return TIMESTAMP_BASE + TS_I2SE_LOCAL_WALCLK(index);
+#else
+	return TIMESTAMP_BASE + TS_I2S_LOCAL_WALCLK(index)
+#endif
+}
+
+static inline uint32_t ssp_ts_tscc_addr(int index)
+{
+#if CONFIG_APOLLOLAKE
+	if (index < DAI_NUM_SSP_BASE)
+		return TIMESTAMP_BASE + TS_I2S_TSCC(index);
+	else
+		return TIMESTAMP_BASE + TS_I2SE_TSCC(index);
+#else
+	return TIMESTAMP_BASE + TS_I2S_TSCC(index)
+#endif
+}
+
+static inline uint64_t io_reg_read_64(uint32_t addr)
+{
+	return (uint64_t)io_reg_read(addr) +
+		(((uint64_t)io_reg_read(addr + 4)) << 32);
+}
+
+static int ssp_ts_config(struct dai *dai, struct sof_timestamp_cfg *cfg)
+{
+	/* Reserved, no settings needed currently */
+	trace_ssp("ssp_ts_config(): SSP=%d, chan=%d",
+		  cfg->index, cfg->dma_chan_index);
+
+	if (cfg->type != SOF_DAI_INTEL_SSP) {
+		trace_ssp_error("ssp_ts_config(): Illegal DAI type");
+		return -EINVAL;
+	}
+
+	if (cfg->index > DAI_NUM_SSP_BASE + DAI_NUM_SSP_EXT - 1) {
+		trace_ssp_error("ssp_ts_config(): ");
+		return -EINVAL;
+	}
+
+	/* TODO, use some existing platform macro */
+	cfg->walclk_rate = 19200000;
+	return 0;
+}
+
+static int ssp_ts_start(struct dai *dai, struct sof_timestamp_cfg *cfg)
+{
+	uint32_t cdmas;
+	uint32_t addr = ssp_ts_local_tsctrl_addr(cfg->index);
+
+	/* Set SSP timestamp registers */
+	spin_lock(dai->lock);
+
+	/* First set CDMAS to DMA channel that is used by this SSP */
+	cdmas = TS_LOCAL_TSCTRL_CDMAS(cfg->dma_chan_index);
+	io_reg_write(addr, cdmas);
+
+	/* Request on demand timestamp */
+	//io_reg_write(addr, TS_LOCAL_TSCTRL_SIP_BIT | cdmas); /* Broken? */
+	//io_reg_write(addr, TS_LOCAL_TSCTRL_ODTS_BIT | cdmas);
+	io_reg_write(addr, TS_LOCAL_TSCTRL_ODTS_BIT);
+	//io_reg_write(addr, TS_LOCAL_TSCTRL_HHTSE_BIT | cdmas);
+
+	spin_unlock(dai->lock);
+	return 0;
+}
+
+static uint64_t ssp_ts_get(struct dai *dai, struct sof_timestamp_cfg *cfg)
+{
+	uint64_t walclk = 0;
+	//uint64_t sample;
+	//uint64_t tscc;
+	//uint32_t offs;
+	uint32_t ntk;
+	uint32_t addr;
+	//int offs_clk;
+	uint32_t tsctrl = ssp_ts_local_tsctrl_addr(cfg->index);
+
+	/* Read SSP timestamp registers */
+	spin_lock(dai->lock);
+	ntk = io_reg_read(tsctrl) & TS_LOCAL_TSCTRL_NTK_BIT;
+	if (ntk) {
+		/* Wall clock */
+		addr = ssp_ts_local_walclk_addr(cfg->index);
+		walclk = io_reg_read_64(addr);
+
+#if 0
+		/* Offset */
+		addr = ssp_ts_local_offs_addr(cfg->index);
+		offs = io_reg_read(addr);
+
+		/* Sample */
+		addr = ssp_ts_local_sample_addr(cfg->index);
+		sample = io_reg_read_64(addr);
+
+		/* TSCC */
+		addr = ssp_ts_tscc_addr(cfg->index);
+		tscc = io_reg_read_64(addr);
+#endif
+
+		/* Clear NTK to enable successive timestamps */
+		io_reg_write(tsctrl, TS_LOCAL_TSCTRL_NTK_BIT);
+	}
+
+	spin_unlock(dai->lock);
+
+	if (ntk) {
+#if 0
+		offs_clk = offs & MASK(15, 0); /* FRM remains zero SSP */
```

Comment:
I presume this is a draft for now? I don't think we should have commented out code like this except temporarily.

Reply:
Yes, that's correct. Use of these features is still open so I didn't remove the code. The DAI / FW arch for this change is the beef and why I want to get the feedback.

Reply:
What should happen with DAIs which simply don't have a hardware timestamp? (I'm not aware of any for the ESAI for example -- unless I missed something in the manual this one simply doesn't have a timestamp at all)

Reply:
Good question, probably the xxx_ts_config() in the driver should return -EINVAL. Also I haven't tested what if ts_ops is NULL. I will try such scenarios with BYT platform.